app.js

```diff
@@ -225,109 +284,101 @@ module.exports = {
-          // The buffer was full, pause the queue to stop the writes until we
-          // get a drain event
-          if (q && !q.paused) {
-            q.pause();
-            targetStream.once('drain', function() {
-              q.resume();
-            });
+          // Write the received content to the targetStream
+          const continueWriting = targetStream.write(JSON.stringify(obj.data) + '\n',
+            'utf8',
+            writeFlushed);
+          if (!continueWriting) {
+            // The buffer was full, pause the queue to stop the writes until we
+            // get a drain event
+            if (q && !q.paused) {
+              q.pause();
+              targetStream.once('drain', function() {
+                q.resume();
+              });
+            }
           }
-        }
-      });
-      // For errors we expect, may or may not be fatal
-      addEventListener(listenerErrorIndicator, internalEE, 'error', function(err) {
-        debug('Error ' + JSON.stringify(err));
-        callback(err);
-      });
-      addEventListener(listenerErrorIndicator, internalEE, 'finished', function(obj) {
-        function emitFinished() {
-          debug('Backup complete - written ' + JSON.stringify(obj));
-          const summary = { total: obj.total };
-          ee.emit('finished', summary);
-          if (callback) callback(null, summary);
-        }
-        if (targetStream === process.stdout) {
-          // stdout cannot emit a finish event so use a final write + callback
-          targetStream.write('', 'utf8', emitFinished);
-        } else {
-          // If we're writing to a file, end the writes and register the
-          // emitFinished function for a callback when the file stream's finish
-          // event is emitted.
-          targetStream.end('', 'utf8', emitFinished);
-        }
-      });
-    });
+        });
+        // For errors we expect, may or may not be fatal
+        addEventListener(listenerErrorIndicator, internalEE, 'error', function(err) {
+          debug('Error ' + JSON.stringify(err));
+          callback(err);
+        });
+        addEventListener(listenerErrorIndicator, internalEE, 'finished', function(obj) {
+          function emitFinished() {
+            debug('Backup complete - written ' + JSON.stringify(obj));
+            const summary = { total: obj.total };
+            ee.emit('finished', summary);
+            if (callback) callback(null, summary);
+          }
+          if (targetStream === process.stdout) {
+            // stdout cannot emit a finish event so use a final write + callback
+            targetStream.write('', 'utf8', emitFinished);
+          } else {
+            // If we're writing to a file, end the writes and register the
+            // emitFinished function for a callback when the file stream's finish
+            // event is emitted.
+            targetStream.end('', 'utf8', emitFinished);
+          }
+        });
+      })
+      .catch(callback);
```

Comment:
Do you think this should move up to after the `.then((backupDB) => validateBackupDb(backupDB))`. My reasoning being that errors that happen in the final `then` block where the backup is actually done are being passed directrly to the callbacks.
I don't think there is, but I can't really figure out if there is a circumstance where we'd end up catching an error that we should otherwise let through. Moving it up would just eliminate that risk. I think once we have the pipeline promise here it won't matter anyway because then we definitely will want a catch at the end, so maybe we can just leave it temporarily. WDYT?

Reply:
I'm ok with moving it up, after all `catch` is just a syntax sugar version of `then(null, err)`. I put it at the bottom with the idea that we are going to fix internalBackup / internalRestore next, so this will be single place we are using  `callback` and since we are doing feature branch it's safe to have the final form here.

Let's move it for now, put it down later.

Reply:
Well, that was educational. It's obvious in hindsight, but `catch` doesn't break a promise chain, so just moving it wasn't enough, it should explicitly throw (or `Promise.reject`) after calling `callback`.

Reply:
Sorry, yeah, that should have been obvious. Mixing promises and callbacks, just don't do it 😄 
Anyway, as we said, this is temporary and won't be shipped and it'll be cleaner anyway once we join all the things together.